<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Methods and tools for the analysis of Genome Wide 
    Identity-by-Descent (gwid) mapping data, focusing on testing whether there 
    is a higher occurrence of Identity-By-Descent (IBD) segments around potential causal variants 
    in cases compared to controls, which is crucial for identifying rare 
    variants. To enhance its analytical power, gwid incorporates a Sliding 
    Window Approach, allowing for the detection and analysis of signals from 
    multiple Single Nucleotide Polymorphisms (SNPs).">
<title>Genome-Wide Identity-by-Descent â€¢ gwid</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Genome-Wide Identity-by-Descent">
<meta property="og:description" content="Methods and tools for the analysis of Genome Wide 
    Identity-by-Descent (gwid) mapping data, focusing on testing whether there 
    is a higher occurrence of Identity-By-Descent (IBD) segments around potential causal variants 
    in cases compared to controls, which is crucial for identifying rare 
    variants. To enhance its analytical power, gwid incorporates a Sliding 
    Window Approach, allowing for the detection and analysis of signals from 
    multiple Single Nucleotide Polymorphisms (SNPs).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">gwid</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/my-vignette.html">gwid</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/soroushmdg/gwid/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="genome-wide-identity-by-descent-">Genome-Wide Identity by Descent <img src="reference/figures/hex1.png" width="175" height="200" align="right"><a class="anchor" aria-label="anchor" href="#genome-wide-identity-by-descent-"></a>
</h1></div>
<!-- badges: start -->

<p>gwid is an R-package designed for the analysis of IBD (Identity by Descent) data, to discover rare alleles (susceptibility regions) associated with case-control phenotype. Although Genome Wide Association Studies (GWAS) successfully reveal numerous common variants linked to diseases, they exhibit lack of power to identify rare alleles. To address this limitation, we have developed a pipeline that employs IBD data (output of refined-IBD software). This methodology encompasses a sequential process for analyzing the aforementioned data within isolated populations. The primary objective of this approach is to enhance the sensitivity of variant detection by utilizing information from genetically related individuals, thereby facilitating the identification of causal variants. An overall representation of the pipeline is visually depicted in the following figure.</p>
<div class="figure" style="text-align: center">
<img src="./reference/figures/final-copy-arrow.png" alt="gwid pipeline" width="100%"><p class="caption">
gwid pipeline
</p>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>The <code>gwid</code> package receives four types of inputs: a genotype file, an IBD file, a haplotype file, and phenotype file. The genotype data is derived from the output of the <a href="https://www.bioconductor.org/packages/release/bioc/html/SNPRelate.html" class="external-link">SNPRelate</a> package in the form of a <strong>gds</strong> file. The IBD file takes the form of tabulated data produced by the <a href="https://faculty.washington.edu/browning/refined-ibd.html" class="external-link">Refined IBD</a> software. Haplotype file comes from the output of the <a href="http://faculty.washington.edu/browning/beagle/beagle.html" class="external-link">Beagle</a>, while phenotype data is represented using an R list.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the stable version of <code>gwid</code> from <a href="https://cran.r-project.org/" class="external-link">CRAN</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"gwid"</span><span class="op">)</span></span></code></pre></div>
<p>Also, you can install the development version of <code>gwid</code> from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"soroushmdg/gwid"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>We demonstrated the key functionalities of gwid using the rheumatoid arthritis (RA) GWAS dataset. This dataset consisted of DNA samples collected from 478 individuals diagnosed with rheumatoid arthritis (RA) and a control group of 1,434 individuals without RA. Genotyping was performed using the Illumina Infinium array. All samples were obtained from a genetically homogeneous population in central Wisconsin exhibiting elevated relatedness structure. Because size of data is large, we use <code>pggyback</code> package to upload and download data from github repository.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("piggyback")</span></span>
<span><span class="fu">piggyback</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/piggyback/reference/pb_download.html" class="external-link">pb_download</a></span><span class="op">(</span>repo <span class="op">=</span> <span class="st">"soroushmdg/gwid"</span>,</span>
<span>            tag <span class="op">=</span> <span class="st">"v0.0.1"</span>,</span>
<span>            dest <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ibd_data_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"//chr3.ibd"</span><span class="op">)</span></span>
<span><span class="va">genome_data_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"//chr3.gds"</span><span class="op">)</span></span>
<span><span class="va">phase_data_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"//chr3.vcf"</span><span class="op">)</span></span>
<span><span class="va">case_control_data_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"//case-cont-RA.withmap.Rda"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="input">Input<a class="anchor" aria-label="anchor" href="#input"></a>
</h3>
<p>In this code we explain each input data files individually. <code>case_control</code> is object of class <code>caco</code> that has phenotype information. <code>snp_data_gds</code> object of class <code>gwas</code> read output of <code>SNPRelate</code> package, we use this package because it is very fast and efficient. <code>haplotype_data</code> object of class <code>phase</code> has haplotype data. <code>ibd_data</code> is an object of <code>gwid</code> class that has IBD information.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/soroushmdg/gwid" class="external-link">gwid</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'gwid'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     print, subset</span></span>
<span></span>
<span><span class="co"># case-control data</span></span>
<span><span class="va">case_control</span> <span class="op">&lt;-</span> <span class="fu">gwid</span><span class="fu">::</span><span class="fu"><a href="reference/case_control.html">case_control</a></span><span class="op">(</span>case_control_rda <span class="op">=</span> <span class="va">case_control_data_file</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">case_control</span><span class="op">)</span> <span class="co"># cases and controls group</span></span>
<span><span class="co">#&gt; [1] "cases" "case1" "case2" "cont1" "cont2" "cont3"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">case_control</span><span class="op">)</span> <span class="co"># in here, we only consider cases,cont1,cont2,cont3</span></span>
<span><span class="co">#&gt;       Length Class  Mode     </span></span>
<span><span class="co">#&gt; cases 478    -none- character</span></span>
<span><span class="co">#&gt; case1 178    -none- character</span></span>
<span><span class="co">#&gt; case2 300    -none- character</span></span>
<span><span class="co">#&gt; cont1 477    -none- character</span></span>
<span><span class="co">#&gt; cont2 478    -none- character</span></span>
<span><span class="co">#&gt; cont3 478    -none- character</span></span>
<span><span class="co"># groups in the study</span></span>
<span><span class="va">case_control</span><span class="op">$</span><span class="va">cases</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span> <span class="co"># first three subject names of cases group</span></span>
<span><span class="co">#&gt; [1] "MC.154405@1075678440" "MC.154595@1075642175" "MC.154701@1076254706"</span></span>
<span></span>
<span><span class="co"># read SNP data (use SNPRelate to convert it to gds) and count number of</span></span>
<span><span class="co"># minor alleles</span></span>
<span><span class="va">snp_data_gds</span> <span class="op">&lt;-</span> <span class="fu">gwid</span><span class="fu">::</span><span class="fu"><a href="reference/build_gwas.html">build_gwas</a></span><span class="op">(</span>gds_data <span class="op">=</span> <span class="va">genome_data_file</span>, </span>
<span>                                 caco <span class="op">=</span> <span class="va">case_control</span>, </span>
<span>                                 gwas_generator <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">snp_data_gds</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "gwas"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">snp_data_gds</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "smp.id"   "snp.id"   "snp.pos"  "smp.indx" "smp.snp"  "caco"     "snps"</span></span>
<span><span class="co"># it has information about counts of minor alleles in each location.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">snp_data_gds</span><span class="op">$</span><span class="va">snps</span><span class="op">)</span> </span>
<span><span class="co">#&gt;    snp_pos case_control value</span></span>
<span><span class="co">#&gt; 1:   66894        cases   627</span></span>
<span><span class="co">#&gt; 2:   66894        case1   240</span></span>
<span><span class="co">#&gt; 3:   66894        case2   387</span></span>
<span><span class="co">#&gt; 4:   66894        cont1   639</span></span>
<span><span class="co">#&gt; 5:   66894        cont2   647</span></span>
<span><span class="co">#&gt; 6:   66894        cont3   646</span></span>
<span></span>
<span><span class="co"># read haplotype data (output of beagle)</span></span>
<span><span class="va">haplotype_data</span> <span class="op">&lt;-</span> <span class="fu">gwid</span><span class="fu">::</span><span class="fu"><a href="reference/build_phase.html">build_phase</a></span><span class="op">(</span>phased_vcf <span class="op">=</span> <span class="va">phase_data_file</span>, </span>
<span>                                    caco <span class="op">=</span> <span class="va">case_control</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">haplotype_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "phase"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">haplotype_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Hap.1" "Hap.2"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">haplotype_data</span><span class="op">$</span><span class="va">Hap.1</span><span class="op">)</span> <span class="co"># 22302 SNP and 1911 subjects</span></span>
<span><span class="co">#&gt; [1] 22302  1911</span></span>
<span></span>
<span><span class="co"># read IBD data (output of Refined-IBD)</span></span>
<span><span class="va">ibd_data</span> <span class="op">&lt;-</span> <span class="fu">gwid</span><span class="fu">::</span><span class="fu"><a href="reference/build_gwid.html">build_gwid</a></span><span class="op">(</span>ibd_data <span class="op">=</span> <span class="va">ibd_data_file</span>, </span>
<span>                             gwas <span class="op">=</span> <span class="va">snp_data_gds</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">ibd_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "gwid"</span></span>
<span><span class="va">ibd_data</span><span class="op">$</span><span class="va">ibd</span> <span class="co"># refined IBD output</span></span>
<span><span class="co">#&gt;                              V1 V2                      V3 V4 V5        V6</span></span>
<span><span class="co">#&gt;      1: MC.AMD127769@0123889787  2    MC.160821@1075679055  1  3  32933295</span></span>
<span><span class="co">#&gt;      2: MC.AMD127769@0123889787  1 MC.AMD107154@0123908746  1  3  29995340</span></span>
<span><span class="co">#&gt;      3: MC.AMD127769@0123889787  2    9474283-1-0238040187  1  3  34165785</span></span>
<span><span class="co">#&gt;      4: MC.AMD127769@0123889787  1    MC.159487@1075679208  2  3  21526766</span></span>
<span><span class="co">#&gt;      5:    MC.163045@1082086165  2    MC.160470@1075679095  1  3  11822616</span></span>
<span><span class="co">#&gt;     ---                                                                   </span></span>
<span><span class="co">#&gt; 377560:    1492602-1-0238095971  2    2235472-1-0238095471  2  3 194785443</span></span>
<span><span class="co">#&gt; 377561:    4618455-1-0238095900  2    3848034-1-0238094219  1  3 190235788</span></span>
<span><span class="co">#&gt; 377562:    MC.160332@1075641581  2    9630188-1-0238038787  2  3 184005719</span></span>
<span><span class="co">#&gt; 377563: MC.AMD122238@0124011436  2    MC.159900@1076254946  1  3 181482803</span></span>
<span><span class="co">#&gt; 377564: MC.AMD105910@0123907456  1    7542312-1-0238039298  1  3 182440135</span></span>
<span><span class="co">#&gt;                V7   V8    V9</span></span>
<span><span class="co">#&gt;      1:  34817627 3.26 1.884</span></span>
<span><span class="co">#&gt;      2:  31752607 4.35 1.757</span></span>
<span><span class="co">#&gt;      3:  35898774 6.36 1.733</span></span>
<span><span class="co">#&gt;      4:  23162240 8.71 1.635</span></span>
<span><span class="co">#&gt;      5:  13523010 5.29 1.700</span></span>
<span><span class="co">#&gt;     ---                     </span></span>
<span><span class="co">#&gt; 377560: 196328849 4.92 1.543</span></span>
<span><span class="co">#&gt; 377561: 192423862 7.77 2.188</span></span>
<span><span class="co">#&gt; 377562: 186184328 5.95 2.179</span></span>
<span><span class="co">#&gt; 377563: 184801115 3.58 3.318</span></span>
<span><span class="co">#&gt; 377564: 183972729 3.03 1.533</span></span>
<span><span class="va">ibd_data</span><span class="op">$</span><span class="va">res</span> <span class="co"># count number of IBD for each SNP location</span></span>
<span><span class="co">#&gt;           snp_pos case_control value</span></span>
<span><span class="co">#&gt;      1:     66894        cases    27</span></span>
<span><span class="co">#&gt;      2:     82010        cases    28</span></span>
<span><span class="co">#&gt;      3:     89511        cases    29</span></span>
<span><span class="co">#&gt;      4:    104972        cases    29</span></span>
<span><span class="co">#&gt;      5:    107776        cases    29</span></span>
<span><span class="co">#&gt;     ---                             </span></span>
<span><span class="co">#&gt; 133808: 197687252        cont3    44</span></span>
<span><span class="co">#&gt; 133809: 197701913        cont3    44</span></span>
<span><span class="co">#&gt; 133810: 197744198        cont3    44</span></span>
<span><span class="co">#&gt; 133811: 197762623        cont3    44</span></span>
<span><span class="co">#&gt; 133812: 197833758        cont3    44</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plot-method">
<code>plot</code> method<a class="anchor" aria-label="anchor" href="#plot-method"></a>
</h3>
<p>The <code>plot</code> function can be applied to the <code>gwid</code> class to display the counts of IBD in each Single SNP among both case and control groups. By utilizing the <code>ly=TRUE</code> parameter, the user has the option to transform the plot into a <code>plotly</code> object, facilitating interactive exploration of the entire chromosome or specific regions of interest through the use of <code>snp_start</code> and <code>snp_end</code> parameters. Additionally, the <code>y</code> parameter enables the inclusion of only specific groups of subjects for consideration.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot count of IBD in chromosome 3</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ibd_data</span>,y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cases"</span>,<span class="st">"cont1"</span><span class="op">)</span>,</span>
<span>     ly <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-6-1.png" width="100%"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Further investigate location between 117M and 122M</span></span>
<span><span class="co"># significant number of IBD's in group cases, compare to cont1, cont2 and cont3.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ibd_data</span>,</span>
<span>     y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cases"</span>,<span class="st">"cont1"</span><span class="op">)</span>,</span>
<span>     snp_start <span class="op">=</span> <span class="fl">117026294</span>,</span>
<span>     snp_end <span class="op">=</span> <span class="fl">122613594</span>,</span>
<span>     ly <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-6-2.png" width="100%"></p>
<p>Through the utilization of the <code>fisher_test</code> method, it becomes possible to calculate p-values within chosen regions. These p-values help assess whether there are noteworthy differences in counts between the case and control groups.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_fisher</span> <span class="op">&lt;-</span> <span class="fu">gwid</span><span class="fu">::</span><span class="fu"><a href="reference/fisher_test.html">fisher_test</a></span><span class="op">(</span><span class="va">ibd_data</span>,<span class="va">case_control</span>,</span>
<span>                                  reference <span class="op">=</span> <span class="st">"cases"</span>,</span>
<span>                                  snp_start <span class="op">=</span> <span class="fl">117026294</span>,</span>
<span>                                  snp_end <span class="op">=</span> <span class="fl">122613594</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">model_fisher</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "test_snps"  "data.table" "data.frame"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">model_fisher</span>, </span>
<span>     y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cases"</span>,<span class="st">"cont1"</span><span class="op">)</span>,</span>
<span>     ly <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-7-1.png" width="100%"></p>
<p>You can perform permutation test as follows:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_permutation</span> <span class="op">&lt;-</span> <span class="fu">gwid</span><span class="fu">::</span><span class="fu"><a href="reference/permutation_test.html">permutation_test</a></span><span class="op">(</span><span class="va">ibd_data</span>,gwas <span class="op">=</span> <span class="va">snp_data_gds</span>,</span>
<span>                                            reference <span class="op">=</span> <span class="st">"cases"</span>,</span>
<span>                                            snp_start <span class="op">=</span> <span class="fl">117026294</span>,</span>
<span>                                            snp_end <span class="op">=</span> <span class="fl">122613594</span>,</span>
<span>                                            nperm <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">model_permutation</span><span class="op">)</span></span></code></pre></div>
<p>The <code>haplotype_structure</code> method can be utilized to extract haplotypes from regions that exhibit IBD patterns in a sliding window manner. <code>w</code> is length of sliding window and</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hap_str</span> <span class="op">&lt;-</span> <span class="fu">gwid</span><span class="fu">::</span><span class="fu"><a href="reference/haplotype_structure.html">haplotype_structure</a></span><span class="op">(</span><span class="va">ibd_data</span>,</span>
<span>                                     phase <span class="op">=</span> <span class="va">haplotype_data</span>,</span>
<span>                                     w <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                     snp_start <span class="op">=</span> <span class="fl">117026294</span>,snp_end <span class="op">=</span> <span class="fl">122613594</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">hap_str</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "haplotype_structure" "data.table"          "data.frame"</span></span>
<span></span>
<span><span class="va">hap_str</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">hap_str</span><span class="op">)</span>,size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,<span class="op">]</span> <span class="co"># structures column have haplotype of length w=10 </span></span>
<span><span class="co">#&gt;    case_control   snp_pos window_number                     smp structures</span></span>
<span><span class="co">#&gt; 1:        cont2 121315813           531    8698972-1-0238040711 0010010000</span></span>
<span><span class="co">#&gt; 2:        case2 121663962           615    MC.159618@1075678747 0100000000</span></span>
<span><span class="co">#&gt; 3:        cases 119438608           316 MC.AMD103791@0124011227 0000000000</span></span>
<span><span class="co">#&gt; 4:        cases 119133106           244 MC.AMD114019@0124011012 0000000000</span></span>
<span><span class="co">#&gt; 5:        cont3 121155122           488    MC.158563@1075641406 0110000000</span></span></code></pre></div>
<p>The <code>haplotype_frequency</code> method can be employed to extract the count of these structures, which can then be plotted for each window.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haplo_freq</span> <span class="op">&lt;-</span> <span class="fu">gwid</span><span class="fu">::</span><span class="fu"><a href="reference/haplotype_frequency.html">haplotype_frequency</a></span><span class="op">(</span><span class="va">hap_str</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># plot haplotype counts in first window (nwin=1).</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">haplo_freq</span>,</span>
<span>   y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cases"</span>, <span class="st">"cont1"</span><span class="op">)</span>,</span>
<span>   plot_type <span class="op">=</span> <span class="st">"haplotype_structure_frequency"</span>,</span>
<span>   nwin <span class="op">=</span> <span class="fl">1</span>, type <span class="op">=</span> <span class="st">"version1"</span>,</span>
<span>   ly <span class="op">=</span> <span class="cn">FALSE</span></span>
<span> <span class="op">)</span></span></code></pre></div>
<p><img src="./reference/figures/hap_version1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=gwid" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/soroushmdg/gwid/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/soroushmdg/gwid/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing gwid</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Soroush Mahmoudiandehkordi <br><small class="roles"> Author, maintainer </small>  </li>
<li>Steven J Schrodi <br><small class="roles"> Author </small>  </li>
<li>Mehdi Maadooliat <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/soroushmdg/gwid/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/soroushmdg/gwid/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Soroush Mahmoudiandehkordi, Steven J Schrodi, Mehdi Maadooliat.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
