---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# gwid

<!-- badges: start -->
<!-- badges: end -->

GWID (Genome Wide Identity by Descent) is an R-package designed for the analysis of IBD (Identity by Descent) data, to discover rare alleles associated with case-control phenotype. Although Genome Wide Association Studies (GWAS) successfully reveal numerous common variants linked to diseases, they exhibit lack of power to identify rare alleles. To address this limitation, we have developed a pipeline that employs IBD data (output of refined-IBD software). This methodology encompasses a sequential process for analyzing the aforementioned data within isolated populations. The primary objective of this approach is to enhance the sensitivity of variant detection by utilizing information from genetically related individuals, thereby facilitating the identification of causal variants. An overall representation of the pipeline is visually depicted in the following figure.



```{r, out.width="70%",include=TRUE, fig.align='center', fig.cap=c("gwid pipeline"), echo=FALSE}
knitr::include_graphics("./inst/Figures/final-copy-arrow.png")
``` 

## Usage
The `gwid` package receives four types of inputs: SNP panel information, IBD information, haplotype data, and data concerning subjects categorized as cases and controls. The SNP panel data is derived from the output of the [SNPRelate](https://www.bioconductor.org/packages/release/bioc/html/SNPRelate.html) package in the form of a **gds** file. The IBD data takes the form of tabulated data produced by the [Refined IBD](https://faculty.washington.edu/browning/refined-ibd.html) software. Haplotype data comes from the output of the [Beagle](http://faculty.washington.edu/browning/beagle/beagle.html), while information about case and control subjects is represented using an R list.

## Installation

You can install the development version of `gwid` from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("soroushmdg/gwid")
```

## Example

The following example is for a SNP panel data from the Marshfield Clinic. subjects in case group has Rheumatoid Arthritis (RA).
```{r, echo=FALSE}
# install.packages("piggyback")
library(piggyback)
pb_download(repo = "soroushmdg/gwid",
            tag = "v0.0.1",
            dest = tempdir())
ibd_data <- paste0(tempdir(),"//chr3.ibd")
genome_data <- paste0(tempdir(),"//chr3.gds")
phase_data <- paste0(tempdir(),"//chr3.vcf")
case_control_data <- paste0(tempdir(),"//case-cont-RA.withmap.Rda")
```

This is a basic example which shows you how to solve a common problem:

```{r example}
library(gwid)

# case-control data
caco <- gwid::case_control(case_control_rda = case_control_data)
names(caco) #cases and controls group
summary(caco) # in here, we only consider cases,cont1,cont2,cont3 groups in the study

# read SNP data (use SNPRelate to convert it to gds) and count number of minor alleles  
pieces <- gwid::build_gwas(gds_data = genome_data,caco = caco,gwas_generator = TRUE)
head(pieces$snps)

# read haplotype data (output of beagle)
myphase <- gwid::build_phase(phased_vcf = phase_data,caco = caco)
names(myphase)

# read ibd data (output of refined ibd)
myregion2 <- gwid::build_gwid(ibd_data = ibd_data,gwas = pieces)
myregion2$ibd
myregion2$res # count number of ibd for each SNP location 

# plot count of IBD in chromosome 3
plot(myregion2,y = c("cases","cont1"),ly = FALSE) 

# Further investigate location between 117M and 122M
# significant number of IBD's in group cases, compare to cont1, cont2 and cont3.
plot(myregion2,y = c("cases","cont1"),snp_start = 117026294,snp_end = 122613594,ly = FALSE) 

```

```{r}
model_fisher <- gwid::fisher_test(myregion2,caco,reference = "cases",
                                             snp_start = 117026294,snp_end = 122613594)

plot(model_fisher, y = c("cases","cont1"),ly = FALSE)
```






You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. `devtools::build_readme()` is handy for this. You could also use GitHub Actions to re-render `README.Rmd` every time you push. An example workflow can be found here: <https://github.com/r-lib/actions/tree/v1/examples>.

You can also embed plots, for example:

```{r pressure, echo = FALSE}
#plot(pressure)
```

In that case, don't forget to commit and push the resulting figure files, so they display on GitHub and CRAN.
